File: Pacman-Game-OOP-master 2/EASendMail/Samples_NetDesktop/VB/Mass/AppProject1/TrafficController.vb
Imports System.Threading

Class TrafficController
    Private _currentThreadId As Integer = Thread.CurrentThread.ManagedThreadId
    Private _concurrentConnections As Integer = 0

    Public Sub DecreaseConnection()
        SyncLock Me
            _concurrentConnections -= 1
        End SyncLock
    End Sub

    Private _maximumConnections As Integer = 10

    Public Property MaximumConnections As Integer
        Get
            Return _maximumConnections
        End Get
        Set(ByVal value As Integer)
            _maximumConnections = value
        End Set
    End Property

    Private _maximumMessagesPerMinutes As Integer = 0 ' means unlimited

    Public Property MaximumMessagesPerMinute As Integer
        Get
            Return _maximumMessagesPerMinutes
        End Get
        Set(ByVal value As Integer)
            _maximumMessagesPerMinutes = value
        End Set
    End Property

    Private _isUncommittedConnection As Boolean = False
    Public Function PrepareIncreaseConnection() As Boolean
        If _currentThreadId <> Thread.CurrentThread.ManagedThreadId Then
            Throw New Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.")
        End If

        If _isUncommittedConnection Then
            Throw New Exception("Uncommitted connection is existed, please commit or rollback it at first.")
        End If

        If _maximumConnections = 0 Then
            Return True
        End If

        If _concurrentConnections >= _maximumConnections Then
            Return False
        End If

        _isUncommittedConnection = True
        Return True
    End Function

    Private _isUncommittedMessage As Boolean = False
    Public Function PrepareIncreaseMessage() As Boolean
        If _currentThreadId <> Thread.CurrentThread.ManagedThreadId Then
            Throw New Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.")
        End If

        If _isUncommittedMessage Then
            Throw New Exception("Uncommitted message is existed, please commit or rollback it at first.")
        End If

        If _maximumMessagesPerMinutes = 0 Then
            Return True
        End If

        If _messageDateTimes.Count >= _maximumMessagesPerMinutes Then
            _clearExpiredMessageCount()
        End If

        If _messageDateTimes.Count >= _maximumMessagesPerMinutes Then
            Return False
        End If

        _isUncommittedMessage = True
        Return True
    End Function

    Private Sub _clearExpiredMessageCount()
        Dim expiredTime As DateTime = DateTime.Now.AddMinutes(-60)

        While _messageDateTimes.Peek() > expiredTime
            _messageDateTimes.Dequeue()
        End While
    End Sub

    Public Sub Commit()
        If _currentThreadId <> Thread.CurrentThread.ManagedThreadId Then
            Throw New Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.")
        End If

        If _isUncommittedConnection Then
            _concurrentConnections += 1
            _isUncommittedConnection = False
        End If

        If _isUncommittedMessage Then
            _messageDateTimes.Enqueue(DateTime.Now)
            _isUncommittedMessage = False
        End If
    End Sub

    Public Sub Rollback()
        If _currentThreadId <> Thread.CurrentThread.ManagedThreadId Then
            Throw New Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.")
        End If

        _isUncommittedConnection = False
        _isUncommittedMessage = False
    End Sub

    Public Sub Reset()
        SyncLock Me
            _concurrentConnections = 0
            _messageDateTimes.Clear()
            _isUncommittedConnection = False
            _isUncommittedMessage = False
        End SyncLock
    End Sub

    Private _messageDateTimes As Queue(Of DateTime) = New Queue(Of DateTime)()
End Class

File: Pacman-Game-OOP-master 2/EASendMail/Samples_NetDesktop/C#/Mass/AppProject1/TrafficController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace AppProject1
{
    class TrafficController
    {
        int _currentThreadId = Thread.CurrentThread.ManagedThreadId;
        int _concurrentConnections = 0;
        public void DecreaseConnection()
        {
            lock (this)
            {
                _concurrentConnections--;
            }
        }

        int _maximumConnections = 10;
        public int MaximumConnections
        {
            get { return _maximumConnections; }
            set { _maximumConnections = value; }
        }

        int _maximumMessagesPerMinutes = 0; // means unlimited
        public int MaximumMessagesPerMinute
        {
            get { return _maximumMessagesPerMinutes; }
            set { _maximumMessagesPerMinutes = value; }
        }

        bool _isUncommittedConnection = false;
        public bool PrepareIncreaseConnection()
        {
            if (_currentThreadId != Thread.CurrentThread.ManagedThreadId)
            {
                throw new Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.");
            }

            if (_isUncommittedConnection)
            {
                throw new Exception("Uncommitted connection is existed, please commit or rollback it at first.");
            }
            // unlimited
            if (_maximumConnections == 0)
            {
                return true;
            }

            if (_concurrentConnections >= _maximumConnections)
            {
                return false;
            }

            _isUncommittedConnection = true;
            return true;

        }
        bool _isUncommittedMessage = false;
        public bool PrepareIncreaseMessage()
        {
            if (_currentThreadId != Thread.CurrentThread.ManagedThreadId)
            {
                throw new Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.");
            }

            if (_isUncommittedMessage)
            {
                throw new Exception("Uncommitted message is existed, please commit or rollback it at first.");
            }

            // unlimited and don't increase message counter
            if (_maximumMessagesPerMinutes == 0)
            {
                return true;
            }

            if (_messageDateTimes.Count >= _maximumMessagesPerMinutes)
            {
                // only clear expired message while out of limit.
                _clearExpiredMessageCount();
            }

            if (_messageDateTimes.Count >= _maximumMessagesPerMinutes)
            {
                return false;
            }

            _isUncommittedMessage = true;
            return true;
        }

        void _clearExpiredMessageCount()
        {
            DateTime expiredTime = DateTime.Now.AddMinutes(-60);
            while (_messageDateTimes.Peek() > expiredTime)
            {
                _messageDateTimes.Dequeue();
            }
        }

        public void Commit()
        {
            if (_currentThreadId != Thread.CurrentThread.ManagedThreadId)
            {
                throw new Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.");
            }

            if (_isUncommittedConnection)
            {
                _concurrentConnections++;
                _isUncommittedConnection = false;
            }

            if (_isUncommittedMessage)
            {
                _messageDateTimes.Enqueue(DateTime.Now);
                _isUncommittedMessage = false;
            }
        }

        public void Rollback()
        {
            if (_currentThreadId != Thread.CurrentThread.ManagedThreadId)
            {
                throw new Exception("PrepareIncreaseConnection method is not thread safe, please invoke it in created thread.");
            }

            _isUncommittedConnection = false;
            _isUncommittedMessage = false;
        }

        public void Reset()
        {
            lock (this)
            {
                _concurrentConnections = 0;
                _messageDateTimes.Clear();
                _isUncommittedConnection = false;
                _isUncommittedMessage = false;
            }
        }

        Queue<DateTime> _messageDateTimes = new Queue<DateTime>();
    }
}
